runtime/src/iree/hal/drivers/vulkan/debug_reporter.cc

```diff
@@ -141,6 +151,9 @@ void iree_hal_vulkan_debug_reporter_free(
         reporter->instance, reporter->messenger,
         reporter->allocation_callbacks);
   }
+  if (reporter->check_errors && reporter->did_error) {
+    IREE_ASSERT(0);
```

Comment:
Do any of the other HAL drivers / device APIs have error reporting that we would similarly want to query "after command buffer completion" (i.e. like `glGetError()`)? The tricky thing with the validation layer / debug utils error reporting is that it can trigger anywhere, so just checking for the success of each individual API call when it happens can miss validation errors. I could see more of a case for error plumbing to `iree_hal_device_t` if several APIs would benefit.

Reply:
we really don't want things like glGetError :P 
relying on optionally implemented error handling with potentially large performance cliffs or non-deterministic behavior is a big no no - that we even want this is sketchy (but I understand) and goes against our own best practices - we definitely don't want to bake it into the API such that we need to support it in other backends

one idea would be that we can feed back errors lazily via normal device methods that may fail due to internal errors anyway - as that's the best a user can ever expect. if the reporter took an iree_atomic_int32_t* that was stored on the VkDeviceHandle we could have things like iree_hal_vulkan_native_semaphore_multi_wait check that and return INTERNAL if it's non-zero, consuming it (reset to 0) would allow for scoped error checks (each test in a gtest suite could report errors, etc without tearing down the device). when the reporter was disabled it'd never get set and not impact things (beyond the read of the always-zero field).

Reply:
Oooo interesting idea. I can give that a try and see what the output looks like.

And yeah, glGetError is gross...

Reply:
if we wanted fancy errors we can use the same technique we use on semaphores to create sticky iree_status_t instances - an atomic int64 iree_status_t that we can compare-and-set (and if compare fails free) would let us get the exact vulkan error back to users

Reply:
Got fancy errors with a sticky status:
```
[ RUN      ] CTS/command_buffer_test.FillBuffer_pattern1_size1_offset0_length1/vulkan
[VULKAN] ! Validation Error: [ VUID-VkMappedMemoryRange-size-01390 ] Object 0: handle = 0xe7e6d0000000000f, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0xdd4e6d8b | vkFlushMappedMemoryRanges(): pMemoryRanges[0].size (4) is not a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize (64) and offset + size (0 + 4 = 4) not equal to the memory size (64). The Vulkan spec states: If size is not equal to VK_WHOLE_SIZE, size must either be a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize, or offset plus size must equal the size of memory (https://vulkan.lunarg.com/doc/view/1.3.268.0/windows/1.3-extensions/vkspec.html#VUID-VkMappedMemoryRange-size-01390)
D:\dev\projects\iree\runtime\src\iree\hal\drivers\vulkan\debug_reporter.cc:112: INTERNAL; Vulkan debug message:
Validation Error: [ VUID-VkMappedMemoryRange-size-01390 ] Object 0: handle = 0xe7e6d0000000000f, type = VK_OBJECT_TYPE_DEVICE_MEMORY; | MessageID = 0xdd4e6d8b | vkFlushMappedMemoryRanges(): pMemoryRanges[0].size (4) is not a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize (64) and offset + size (0 + 4 = 4) not equal to the memory size (64). The Vulkan spec states: If size is not equal to VK_WHOLE_SIZE, size must either be a multiple of VkPhysicalDeviceLimits::nonCoherentAtomSize, or offset plus size must equal the size of memory (https://vulkan.lunarg.com/doc/view/1.3.268.0/windows/1.3-extensions/vkspec.html#VUID-VkMappedMemoryRange-size-01390); stack:
  0x00007ff7a8fd8e4c vulkan_command_buffer_test <iree_hal_vulkan_debug_utils_message_callback+0xac> (D:\dev\projects\iree\runtime\src\iree\hal\drivers\vulkan\debug_reporter.cc:109)
  0x00007ffb7828cacd ???
  0x00007ffb7828e3a7 ???
  0x00007ffb77d99235 ???
  0x00007ffb77e2bc84 ???
  0x00007ffb77e2bd6e ???
  0x00007ffb77f67181 ???
  0x00007ff7a8fe77bc vulkan_command_buffer_test <iree_hal_vulkan_native_buffer_flush_range+0x9c> (D:\dev\projects\iree\runtime\src\iree\hal\drivers\vulkan\native_buffer.cc:202)
  0x00007ff7a8fc7c35 vulkan_command_buffer_test <iree_hal_buffer_mapping_flush_range+0x95> (D:\dev\projects\iree\runtime\src\iree\hal\buffer.c:917)
  0x00007ff7a8fc748b vulkan_command_buffer_test <iree_hal_buffer_map_fill+0x20b> (D:\dev\projects\iree\runtime\src\iree\hal\buffer.c:665)
  0x00007ff7a8fc7b8c vulkan_command_buffer_test <iree_hal_buffer_map_zero+0x1c> (D:\dev\projects\iree\runtime\src\iree\hal\buffer.c:580)
  0x00007ff7a8fb0ff9 vulkan_command_buffer_test <iree::hal::cts::command_buffer_test::CreateZeroedDeviceBuffer+0xd9> (D:\dev\projects\iree\runtime\src\iree\hal\cts\command_buffer_test.h:43)
  0x00007ff7a8fb34d4 vulkan_command_buffer_test <iree::hal::cts::command_buffer_test::RunFillBufferTest+0x34> (D:\dev\projects\iree\runtime\src\iree\hal\cts\command_buffer_test.h:57)
  0x00007ff7a8fb852b vulkan_command_buffer_test <iree::hal::cts::command_buffer_test_FillBuffer_pattern1_size1_offset0_length1_Test::TestBody+0x7b> (D:\dev\projects\iree\runtime\src\iree\hal\cts\command_buffer_test.h:270)
  0x00007ff7a903c02b vulkan_command_buffer_test <testing::internal::HandleSehExceptionsInMethodIfSupported<testing::Test,void>+0xb> (D:\dev\projects\iree\third_party\googletest\googletest\src\gtest.cc:2606)

Assertion failed: 0, file D:\dev\projects\iree\runtime\src\iree\hal\drivers\vulkan\debug_reporter.cc, line 204
```

Now figuring out the plumbing to remove the assert...

> if the reporter took an iree_atomic_int32_t* that was stored on the VkDeviceHandle we could have things like iree_hal_vulkan_native_semaphore_multi_wait check that

The messenger is created at the `VkInstance` level (in [`vulkan_driver.cc`](https://github.com/openxla/iree/blob/main/runtime/src/iree/hal/drivers/vulkan/vulkan_driver.cc)), not the `VkDevice` level:
https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkCreateDebugUtilsMessengerEXT.html. I could create a reporter for each device (possibly missing instance-level messages). Would be nice to scope things so errors on each device are isolated. If we accept that this plumbing cuts across a few layers, I could try passing the reporter down into the device and then having a few places (like the semaphore wait) check for errors.

Reply:
yeah, this is the problem with this kind of thing - it's really not meant as a way to get errors back to users :/
if we kept the atomic iree_status_t on the reporter and then let all devices pull from it that's probably fine, as the debug reporter shouldn't be enabled in any release/production builds so there shouldn't be interference